File: clients/desktop/src-tauri/src/deeplink.rs
// oort:// deep link intake (ADR-0133 P2, MOMO-603; scheme renamed in goal B13).
//
// Contract source of truth: `docs/onboarding-deeplink.md`.
//   oort://join?server=<percent-encoded base URL>&code=<invite code>
// Two parameters, order-independent, unknown parameters ignored. This is a
// straight port of the macOS parser (`clients/macOS/Sources/MomoMac/MomoDeepLink.swift`)
// so the two shells prefill identically from the same link — including the parts
// that look like details but are contract: the action may be the authority
// (`oort://join`) or the first path segment (`oort:join`), parameter names are
// matched case-insensitively, the FIRST occurrence of a name wins, and a link
// carrying only one of the two values is still worth surfacing because it still
// saves the person typing.
//
// Deliberately NOT validated here: whether `server` is a usable base URL. The
// macOS side re-validates downstream (`validatedBaseURL()`), and so does the web
// join surface, so a single owner keeps that rule instead of two drifting copies.
//
// Percent-decoding is done by hand rather than through `Url::query_pairs()`,
// because that applies `application/x-www-form-urlencoded` rules and would turn a
// literal `+` into a space. The contract is RFC 3986 percent-encoding, which the
// macOS `URLComponents` path follows, so `+` must survive as `+`.

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Mutex;

use percent_encoding::percent_decode_str;
use serde::Serialize;
use tauri::{AppHandle, Emitter, Runtime, State};
use url::Url;

/// Emitted to the webview for every accepted join link.
///
/// This is a Tauri IPC channel name, not a URL scheme, and it is NOT part of the
/// momo -> oort rebrand: renaming it would only have to be renamed in lockstep
/// in `lib/tauri.ts` for no one's benefit. Nobody outside the shell sees it.
pub const DEEP_LINK_EVENT: &str = "momo:deep-link";

/// The scheme links are minted with (goal B13, momo -> oort).
const JOIN_SCHEME: &str = "oort";

/// Every scheme a link is accepted under. The old name is absorbed rather than
/// dropped: an invite link lives in someone's inbox for days after it is sent,
/// and it was correct when it was sent. `tauri.conf.json` registers both with
/// the OS so both actually reach this parser.
const ACCEPTED_SCHEMES: [&str; 2] = [JOIN_SCHEME, "momo"];
const JOIN_ACTION: &str = "join";

/// A join link the webview can prefill from.
///
/// `server` and `code` are percent-decoded and trimmed; either may be empty, but
/// never both (a link with nothing usable is dropped rather than delivered).
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct JoinLink {
    /// The raw link as received, for logging and for the "opened from a link"
    /// affordance. Carries the invite code, so it is never logged at info level.
    pub url: String,
    pub server: String,
    pub code: String,
}

/// Parses a join link under any accepted scheme, or `None` for anything else.
pub fn parse_join(url: &Url) -> Option<JoinLink> {
    if !ACCEPTED_SCHEMES.contains(&url.scheme()) {
        return None;
    }
    if resolved_action(url)? != JOIN_ACTION {
        return None;
    }

    let mut server: Option<String> = None;
    let mut code: Option<String> = None;
    for (name, value) in query_pairs(url.query().unwrap_or("")) {
        match name.to_ascii_lowercase().as_str() {
            "server" if server.is_none() => server = Some(value),
            "code" if code.is_none() => code = Some(value),
            _ => {} // Unknown parameters are ignored per the shared contract.
        }
    }

    let server = server.unwrap_or_default().trim().to_string();
    let code = code.unwrap_or_default().trim().to_string();
    if server.is_empty() && code.is_empty() {
        return None;
    }
    Some(JoinLink {
        url: url.as_str().to_string(),
        server,
        code,
    })
}

/// The action is the authority (`oort://join`) or, when the link omits the
/// authority (`oort:join`), the first non-empty path segment. Case-insensitive.
fn resolved_action(url: &Url) -> Option<String> {
    if let Some(host) = url.host_str() {
        if !host.is_empty() {
            return Some(host.to_ascii_lowercase());
        }
    }
    url.path()
        .split('/')
        .find(|segment| !segment.is_empty())
        .map(|segment| {
            percent_decode_str(segment)
                .decode_utf8_lossy()
                .to_ascii_lowercase()
        })
}

/// RFC 3986 percent-decoding of a raw query string. No `+` to space.
fn query_pairs(query: &str) -> Vec<(String, String)> {
    query
        .split('&')
        .filter(|pair| !pair.is_empty())
        .map(|pair| {
            let (name, value) = pair.split_once('=').unwrap_or((pair, ""));
            (decode(name), decode(value))
        })
        .collect()
}

fn decode(value: &str) -> String {
    percent_decode_str(value).decode_utf8_lossy().into_owned()
}

/// Buffers links that arrive before the webview can hear them.
///
/// macOS delivers the launch URL while the app is still starting, well before
/// React has mounted and subscribed, so a pure event bridge would drop exactly
/// the case that matters most: clicking an invite link with the app closed. The
/// handshake is one-shot — `deep_link_take_pending` marks the webview ready and
/// drains the buffer, and everything after that is delivered as events only, so
/// a link is never both replayed and emitted.
#[derive(Default)]
pub struct DeepLinkState {
    pending: Mutex<Vec<JoinLink>>,
    webview_ready: AtomicBool,
}

impl DeepLinkState {
    /// Route one incoming link: emit it if the webview is listening, buffer it
    /// otherwise. Links that are not a join link are dropped silently.
    pub fn deliver<R: Runtime>(&self, app: &AppHandle<R>, url: &Url) {
        let Some(link) = parse_join(url) else {
            return;
        };
        if self.webview_ready.load(Ordering::SeqCst) {
            let _ = app.emit(DEEP_LINK_EVENT, &link);
        } else if let Ok(mut pending) = self.pending.lock() {
            pending.push(link);
        }
    }
}

/// Drains links that arrived before the webview subscribed, and marks it ready.
///
/// Call once, immediately after attaching the `momo:deep-link` listener.
#[tauri::command]
pub fn deep_link_take_pending(state: State<'_, DeepLinkState>) -> Vec<JoinLink> {
    state.webview_ready.store(true, Ordering::SeqCst);
    state
        .pending
        .lock()
        .map(|mut pending| std::mem::take(&mut *pending))
        .unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse(raw: &str) -> Option<JoinLink> {
        parse_join(&Url::parse(raw).expect("test link parses as a URL"))
    }

    #[test]
    fn parses_the_canonical_link() {
        let link = parse("oort://join?server=http%3A%2F%2FMacBook.local%3A28000&code=abc-123")
            .expect("canonical link is accepted");
        assert_eq!(link.server, "http://MacBook.local:28000");
        assert_eq!(link.code, "abc-123");
    }

    #[test]
    fn parameter_order_does_not_matter() {
        let reversed = parse("oort://join?code=abc-123&server=https%3A%2F%2Fapi.example.com")
            .expect("reversed order is accepted");
        assert_eq!(reversed.server, "https://api.example.com");
        assert_eq!(reversed.code, "abc-123");
    }

    #[test]
    fn unknown_parameters_are_ignored() {
        let link = parse("oort://join?utm=mail&code=abc&server=https%3A%2F%2Fa.example&x=1")
            .expect("extra parameters do not invalidate the link");
        assert_eq!(link.server, "https://a.example");
        assert_eq!(link.code, "abc");
    }

    #[test]
    fn accepts_the_authority_free_form() {
        assert!(parse("oort:join?code=abc").is_some());
    }

    #[test]
    fn rejects_other_schemes_and_actions() {
        assert!(parse("https://join?code=abc").is_none());
        assert!(parse("oort://invite?code=abc").is_none());
        assert!(parse("slack://join?code=abc").is_none());
    }

    /// **The old scheme still opens.**
    ///
    /// goal B13 renamed the minted scheme `momo://` -> `oort://`, and dropping
    /// the old one would have broken every invite already sitting in somebody's
    /// inbox — links that were correct when they were sent. Both are registered
    /// with the OS (`tauri.conf.json`) so both actually arrive here, and this
    /// test is what stops the second entry from being "tidied up" later.
    #[test]
    fn still_opens_a_link_minted_under_the_old_scheme() {
        let link = parse("momo://join?server=https%3A%2F%2Fa.example&code=abc")
            .expect("a momo:// invite from before the rename");
        assert_eq!(link.server, "https://a.example");
        assert_eq!(link.code, "abc");
        assert!(parse("momo:join?code=abc").is_some());
    }

    #[test]
    fn rejects_a_link_with_nothing_to_prefill() {
        assert!(parse("oort://join").is_none());
        assert!(parse("oort://join?server=&code=").is_none());
        assert!(parse("oort://join?utm=mail").is_none());
    }

    #[test]
    fn keeps_a_partial_link_because_it_still_saves_typing() {
        let server_only = parse("oort://join?server=https%3A%2F%2Fa.example").expect("server only");
        assert_eq!(server_only.code, "");
        let code_only = parse("oort://join?code=abc").expect("code only");
        assert_eq!(code_only.server, "");
    }

    #[test]
    fn a_plus_in_a_value_is_not_a_space() {
        // base64url codes never contain '+', but form-urlencoded decoding would
        // silently corrupt one that did; RFC 3986 is the contract.
        let link = parse("oort://join?code=a%2Bb").expect("link parses");
        assert_eq!(link.code, "a+b");
    }

    #[test]
    fn parameter_names_are_case_insensitive_and_first_wins() {
        let link =
            parse("oort://join?SERVER=https%3A%2F%2Fa.example&server=https%3A%2F%2Fb.example")
                .expect("link parses");
        assert_eq!(link.server, "https://a.example");
    }
}

File: packages/momo-core/src/features/auth/deepLink.ts
import { normalizeServerUrl } from "../../lib/serverUrl";
// -----------------------------------------------------------------------------
// 실기기 제약 (스파이크 #837 게이트 2) — **RN 호스트는 `react-native-url-polyfill`
// 을 먼저 설치해야 한다.**
//
// RN 코어의 `URL` 은 정규식 기반이고 WHATWG 구현이 아니라 `oort://join?...` 같은
// 커스텀 스킴을 파싱하지 못한다: 같은 케이스 19건이 RN 코어 URL 에서 19/19 실패,
// `react-native-url-polyfill@4.0.0` 에서 19/19 통과했다.
// **이 파일 자체는 무수정으로 통과했다** — 고칠 것은 호스트의 부트스트랩이다.
// -----------------------------------------------------------------------------



// =============================================================================
// Invite deep link (`oort://join`) — the web consumer of the contract in
// docs/onboarding-deeplink.md, which the ops CLI (MOMO-584) and the mac client
// (MomoDeepLinkParser, MOMO-585) already implement byte for byte:
//
//     oort://join?server=<percent-encoded base URL>&code=<invite code>
//
// goal B13: the scheme was `momo://` through MOMO-584/585. The product is oort
// now, so links are MINTED as `oort://` and still ACCEPTED as either — see
// ACCEPTED_SCHEMES. The repo, the crates and the binaries keep the momo
// codename; only the surface a person sees moved.
//
//   - exactly two meaningful parameters, order-independent;
//   - unknown parameters ignored (forward compatibility);
//   - `server` arrives percent-encoded and is re-validated here, because a link
//     is untrusted input no matter who forwarded it.
//
// Two ways in, one parser:
//   desktop  the shell hands over the URL (`momo:deep-link`, lib/tauri.ts);
//   browser  there is no custom scheme, so the same parameters ride the page
//            URL instead: `?server=...&code=...`, `?code=...` alone, or
//            `?join=<percent-encoded oort:// link>` when a whole link was
//            pasted. The HashRouter's own query (`#/path?code=...`) is read too,
//            since that is where a link lands once the router owns the URL.
//
// The invite code is a bearer secret. It is parsed here and handed to the form;
// it is never logged, and the connect screen strips it from the address bar
// once it has been read.
// =============================================================================

export interface JoinPrefill {
  /** Validated API base, or "" when the link only carried a code. */
  serverUrl: string;
  /** Invite code, or "" when the link only carried a server. */
  inviteCode: string;
}

/**
 * The scheme links are MINTED with (goal B13, momo -> oort rebrand).
 * `buildJoinLink` and the ops CLI both emit this one.
 */
export const JOIN_SCHEME = "oort";

/**
 * Every scheme a link is ACCEPTED under.
 *
 * The old name is absorbed rather than ignored, and that is a decision about
 * links that already exist: an invite is handed over out of band — email, Slack,
 * a message — and it stays valid for days. Dropping `momo://` would break links
 * that were correct when they were sent, to no one's benefit, so the OS keeps
 * both registrations (`tauri.conf.json`, both `Info.plist`s) and every consumer
 * keeps both here.
 */
const ACCEPTED_SCHEMES = [JOIN_SCHEME, "momo"] as const;
const JOIN_ACTION = "join";
const LINK_ACTION = "link";

function schemeAction(url: URL): string {
  return url.hostname !== ""
    ? url.hostname.toLowerCase()
    : (url.pathname.split("/").filter(Boolean)[0] ?? "").toLowerCase();
}

function parseAcceptedScheme(raw: string): URL | null {
  let url: URL;
  try {
    url = new URL(raw.trim());
  } catch {
    return null;
  }
  const protocol = url.protocol.toLowerCase();
  if (!ACCEPTED_SCHEMES.some((scheme) => protocol === `${scheme}:`)) return null;
  return url;
}

function firstParam(params: URLSearchParams, name: string): string {
  for (const [key, value] of params) {
    if (key.toLowerCase() === name) return value.trim();
  }
  return "";
}

/** A link's `server` is only accepted once it validates as a base URL. */
function validatedServer(raw: string): string {
  if (raw === "") return "";
  const checked = normalizeServerUrl(raw);
  return checked.ok ? checked.base : "";
}

function prefillFrom(params: URLSearchParams): JoinPrefill | null {
  const serverUrl = validatedServer(firstParam(params, "server"));
  const inviteCode = firstParam(params, "code");
  if (serverUrl === "" && inviteCode === "") return null;
  return { serverUrl, inviteCode };
}

/**
 * Parse a join link under any accepted scheme. Returns null when the URL is not
 * a join link or
 * carries nothing usable, so a stray link is ignored rather than half-applied.
 *
 * The action is the authority (`oort://join`) or, when the link omits it
 * (`oort:join`), the first path segment. Same rule as the mac parser.
 */
export function parseJoinDeepLink(raw: string): JoinPrefill | null {
  const url = parseAcceptedScheme(raw);
  if (!url) return null;
  if (schemeAction(url) !== JOIN_ACTION) return null;
  return prefillFrom(url.searchParams);
}

/**
 * Browser fallback: the same two parameters off the page URL. Reads the real
 * query first, then the hash query the HashRouter leaves behind, and unwraps a
 * whole `oort://join` link passed as `?join=`.
 */
export function parseJoinFromPageUrl(href: string): JoinPrefill | null {
  let url: URL;
  try {
    url = new URL(href);
  } catch {
    return null;
  }

  const hashQueryIndex = url.hash.indexOf("?");
  const hashParams = new URLSearchParams(
    hashQueryIndex >= 0 ? url.hash.slice(hashQueryIndex + 1) : ""
  );

  for (const params of [url.searchParams, hashParams]) {
    const wrapped = firstParam(params, "join");
    if (wrapped !== "") {
      const parsed = parseJoinDeepLink(wrapped);
      if (parsed) return parsed;
    }
    const direct = prefillFrom(params);
    if (direct) return direct;
  }
  return null;
}

/** The parameters the connect screen strips once a prefill has been read. */
export const JOIN_URL_PARAMS = ["join", "server", "code"] as const;

/**
 * Remove the invite parameters from a URL string, preserving everything else.
 * The code is a bearer secret: leaving it in the address bar puts it in history
 * and in the next screenshot.
 */
export function urlWithoutJoinParams(href: string): string {
  let url: URL;
  try {
    url = new URL(href);
  } catch {
    return href;
  }
  for (const name of JOIN_URL_PARAMS) url.searchParams.delete(name);

  const hashQueryIndex = url.hash.indexOf("?");
  if (hashQueryIndex >= 0) {
    const path = url.hash.slice(0, hashQueryIndex);
    const hashParams = new URLSearchParams(url.hash.slice(hashQueryIndex + 1));
    for (const name of JOIN_URL_PARAMS) hashParams.delete(name);
    const rest = hashParams.toString();
    url.hash = rest === "" ? path : `${path}?${rest}`;
  }
  return url.toString();
}

/**
 * Device-link deep link (`oort://link?server=…&token=…`, ADR-0180 D2).
 *
 * Same grammar as `join`: two meaningful parameters, order-independent,
 * unknown parameters ignored, `momo://` absorbed. Both `server` and `token`
 * are required; an unusable `server` rejects the whole link (unlike join,
 * which can prefill a code alone).
 */
export interface DeviceLinkPrefill {
  /** Validated API base. */
  serverUrl: string;
  /** Raw 32-byte base64url voucher (43 characters). */
  token: string;
}

/**
 * True when the URL is a device-link *action* (`oort://link` / `momo://link`),
 * even if `server` or `token` is unusable. The connect screen uses this to
 * tell "malformed link" from "not our URL".
 */
export function isDeviceLinkAction(raw: string): boolean {
  const url = parseAcceptedScheme(raw);
  return url !== null && schemeAction(url) === LINK_ACTION;
}

function deviceLinkPrefillFrom(params: URLSearchParams): DeviceLinkPrefill | null {
  const serverUrl = validatedServer(firstParam(params, "server"));
  const token = firstParam(params, "token");
  if (serverUrl === "" || token === "") return null;
  return { serverUrl, token };
}

/**
 * Parse a device-link URL. Null when it is not one, is missing a parameter,
 * or carries a `server` that does not validate as a base URL.
 */
export function parseDeviceLinkDeepLink(raw: string): DeviceLinkPrefill | null {
  const url = parseAcceptedScheme(raw);
  if (!url) return null;
  if (schemeAction(url) !== LINK_ACTION) return null;
  return deviceLinkPrefillFrom(url.searchParams);
}
